### ISSUE TITLE
Signal handling for epoll_wait can leak processes

### ISSUE THREAD

**Issue by bturner** [brettwooldridge/NuProcess#124]
Unexpected signal delivery to the JVM while NuProcess's `ProcessEpoll` is blocked in `epoll_wait` can result in it leaking zombie processes.

Per `signal(7)`:
```
       The following interfaces are never restarted after being interrupted by
       a signal handler, regardless of the use of SA_RESTART; they always fail
       with the error EINTR when interrupted by a signal handler:

       * File     descriptor     multiplexing    interfaces:    epoll_wait(2),
         epoll_pwait(2), poll(2), ppoll(2), select(2), and pselect(2).
```

This means even if the JVM registers signal handlers with `SA_RESTART`, `epoll_wait` will not be restarted.

Looking through JDK's native code at places where it uses `epoll_wait`, it wraps them in a `RESTARTABLE` macro ([defined here](https://github.com/JetBrains/jdk8u_jdk/blob/94318f9185757cc33d2b8d527d36be26ac6b7582/src/solaris/native/sun/nio/ch/nio_util.h#L33-L37)) which detects a `-1` return paired with `EINTR` and re-runs `epoll_wait`. `ProcessEpoll` should have something similar.

Debugging this leak was quite challenging due to NuProcess's (lack of) error handling.
* `ProcessEpoll` [throws a `RuntimeException`](https://github.com/brettwooldridge/NuProcess/blob/master/src/main/java/com/zaxxer/nuprocess/linux/ProcessEpoll.java#L221) with a minimal message and no `errno` to indicate what the error was, but that doesn't really matter because
* `BaseEventProcessor.run` [catches the exception and aborts](https://github.com/brettwooldridge/NuProcess/blob/master/src/main/java/com/zaxxer/nuprocess/internal/BaseEventProcessor.java#L84-L87) without any logging

I ended up [expanding the exception message](https://github.com/atlassian/NuProcess/commit/3ef3e06caf22c48dde81a11c4e2428cf2b0d8780#diff-4e5c48d2691d693151c8dde51faadc114e5d53003ecfe75b686736600398405fR236) and [adding some logging](https://github.com/atlassian/NuProcess/commit/3ef3e06caf22c48dde81a11c4e2428cf2b0d8780#diff-b466e66bd6d8b8c2049aebaef7e9a1a9a433ade7d5a37e89eef64f6dbd850161R94) to finally get to the bottom of it.

**Comment by bturner** [comment]
If the `stdout` or `stderr` buffers fill up, NuProcess also ends up leaking those as zombies due to [this check](https://github.com/brettwooldridge/NuProcess/blob/master/src/main/java/com/zaxxer/nuprocess/internal/BasePosixProcess.java#L370-L375) (which triggers the same log-less abort in `BaseEventProcessor`).

There's documentation that those buffers shouldn't be allowed to fill, and I've got code that prevents it--except in one case: Where I'm in an `onStdout` (or `onStderr`) callback and I call `NuProcess.destroy`. In that case, I (wrongly) assumed the state of the buffer wouldn't matter, but it still does. I've since updated my handlers to purge the buffers after calling `destroy` to prevent zombies.
